in-progress/8025-publishing-tx-objects/design.md

```diff
@@ -398,6 +398,104 @@ In the event of a "failed" transaction, the transaction will appear in the block
+
+### The Request Response Protocol
+
+#### Protocol Identification 
+Each request response interaction is identified by a protocolID (a.k.a. SubProtocol). 
+This protocol will take a form similar to the `gossipSub` protocol topic, the form is expected to be as follows. 
+
+```
+/aztec/req/{subprotocol}/{version}
+```
+
+Where subprotocol details the type of request being made, and version details the version of the protocol being used, the exact form of version is yet to be determined. It may represent a semantic version (such as a hardfork), or it may represent a specific protocol implementation version (semver identifier). 
+
+Whatever version standard is used will match that used by the `gossipSub` protocol topics.
+
+#### Request / Response Interactions
+When a validator requires a payload from a peer, they will dial their peer's, opening up ONE stream per interaction. This stream will be closed when this interaction is complete, regardless of whether the interaction was successful or not.
+
+#### Request / Response Interfaces
+
+Each request response interaction is explicitly defined by the protocolID (a.k.a. SubProtocol). This protocolID can be used to determine the expected message request types, and the expected data response types.
+
+*Request / Response Mapping*
+
+```ts
+/**
+ * The Request Response Pair interface defines the methods that each
+ * request response pair must implement
+ */
+interface RequestResponsePair<Req, Res> {
+  request: new (...args: any[]) => Req;
+  /**
+   * The response must implement the static fromBuffer method (generic serialisation)
+   */
+  response: {
+    new (...args: any[]): Res;
+    fromBuffer(buffer: Buffer): Res;
+  };
+}
+```
+
+For an example transaction request protocol, the serialized request and response types are defined as follows:
+- The request will send a TransactionHash, which means the responder should expect strict serialization of the TransactionHash type.
+- The response will send a Transaction, which means the responder should expect strict serialization of the Transaction type.
+
+```ts
+  '/aztec/req/tx/0.1.0': {
+    request: TxHash,
+    response: Tx,
+  },
+```
+
+#### Making a request 
+When opening a stream with a peer, a specific protocolID (a.k.a. SubProtocol) will be used to determine what data the peer requires. For example, a SubProtocol with id `/aztec/req/tx/0.1.0` represents a stream that will be used to request a transaction payload from a peer.
```

Comment:
Just curious, how does req/resp pick a peer to query?

Reply:
Right now it round robins through it's peer list, but I beleive spyros is working on a peer ranking system, so it should choose the first peer based on its ranking